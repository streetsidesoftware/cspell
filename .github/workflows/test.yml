name: Test
on:
  pull_request:
  push:
    branches:
      - master
      - cspell4

jobs:
  test:

    runs-on: ${{ matrix.os }}

    strategy:
      matrix:
        node-version:
          - 10.x
          - 12.x
          - 14.x

        os:
          - ubuntu-latest
          - windows-latest
          - macos-latest

    steps:
    - uses: actions/checkout@v2

    - name: Use Node.js ${{ matrix.node-version }}
      uses: actions/setup-node@v2.1.2
      with:
        node-version: ${{ matrix.node-version }}

    - run: npm ci
    - run: npm test
    - run: node ./bin.js trace test
    - run: node ./bin.js "**/{*.ts,*.md}"
