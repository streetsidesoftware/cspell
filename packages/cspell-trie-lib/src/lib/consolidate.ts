import { trieNodeToRoot } from './TrieNode/trie-util.js';
import type { TrieNode, TrieRoot } from './TrieNode/TrieNode.js';
import { FLAG_WORD } from './TrieNode/TrieNode.js';

/**
 * Consolidate to DAWG
 * @param root the root of the Trie tree
 */
export function consolidate(root: TrieRoot): TrieRoot {
    let count = 0;
    const signatures = new Map<string, TrieNode>();
    const cached = new Map<TrieNode, number>();
    const knownMap = new Map<TrieNode, TrieNode>();

    function signature(n: TrieNode): string {
        const isWord = n.f ? '*' : '';
        const ref = n.c ? JSON.stringify(Object.entries(n.c).map(([k, n]) => [k, cached.get(n)])) : '';
        return isWord + ref;
    }

    function findEow(n: TrieNode): TrieNode | undefined {
        if (n.f && !n.c) return n;
        let r: TrieNode | undefined;
        // istanbul ignore else
        if (n.c) {
            for (const c of Object.values(n.c)) {
                r = findEow(c);
                // istanbul ignore else
                if (r) break;
            }
        }
        return r;
    }

    function compareMaps(a: [string, TrieNode][], b: Record<string, TrieNode>): boolean {
        for (const e of a) {
            if (b[e[0]] !== e[1]) return false;
        }
        return a.length === b.size;
    }

    function deepCopy(n: TrieNode): TrieNode {
        const k = knownMap.get(n);
        if (k) {
            return k;
        }

        const orig = n;
        if (n.c) {
            const children = Object.entries(n.c).map((c) => [c[0], deepCopy(c[1])] as [string, TrieNode]);
            if (!compareMaps(children, n.c)) {
                n = { f: n.f, c: Object.fromEntries(children) };
            }
        }
        const sig = signature(n);
        const ref = signatures.get(sig);
        if (ref) {
            knownMap.set(orig, ref);
            return ref;
        }
        Object.freeze(n);
        signatures.set(sig, n);
        cached.set(n, count++);
        knownMap.set(orig, n);
        return n;
    }

    function process(n: TrieNode): TrieNode {
        if (cached.has(n)) {
            return n;
        }
        if (Object.isFrozen(n)) {
            return knownMap.get(n) || deepCopy(n);
        }
        if (n.c) {
            const children = Object.entries(n.c)
                .sort((a, b) => (a[0] < b[0] ? -1 : 1))
                .map((c) => [c[0], process(c[1])] as [string, TrieNode]);
            n.c = Object.fromEntries(children);
        }
        const sig = signature(n);
        const ref = signatures.get(sig);
        if (ref) {
            return ref;
        }
        signatures.set(sig, n);
        cached.set(n, count++);
        return n;
    }

    // Add end of word to the set of signatures and cache it.
    const eow = findEow(root) || { f: FLAG_WORD, c: undefined };
    signatures.set(signature(eow), eow);
    cached.set(eow, count++);
    return trieNodeToRoot(process(root), root);
}
