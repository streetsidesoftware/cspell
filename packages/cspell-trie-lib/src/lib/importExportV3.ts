import { TrieNode, FLAG_WORD, TrieRoot } from './TrieNode';
import { Sequence, genSequence } from 'gensequence';
import { bufferLines } from './bufferLines';
import { trieNodeToRoot } from './util';

const EOW = '$';
const BACK = '<';
const EOL = '\n';
const LF = '\r';
const REF = '#';
const EOR = ';';
const ESCAPE = '\\';

const specialCharacters = new Set(
    [EOW, BACK, EOL, REF, EOR, ESCAPE, LF]
        .concat('0123456789'.split(''))
        .concat('`~!@#$%^&*()_-+=[]{};:\'"<>,./?\\|'.split(''))
);

const specialCharacterMap = new Map([
    ['\n', '\\n'],
    ['\r', '\\r'],
    ['\\', '\\\\'],
]);
const characterMap = new Map([...specialCharacterMap].map((a) => [a[1], a[0]]));

export const DATA = '__DATA__';

function generateHeader(base: number, comment: string): Sequence<string> {
    const header = ['#!/usr/bin/env cspell-trie reader', 'TrieXv3', 'base=' + base]
        .concat(comment ? comment.split('\n').map((a) => '# ' + a) : [])
        .concat(['# Data:', DATA]);
    return genSequence(header).map((a) => a + '\n');
}

export interface ExportOptions {
    base?: number;
    comment?: string;
}

/**
 * Serialize a TrieRoot.
 */
export function serializeTrie(root: TrieRoot, options: ExportOptions | number = 16): Sequence<string> {
    options = typeof options === 'number' ? { base: options } : options;
    const { base = 16, comment = '' } = options;
    const radix = base > 36 ? 36 : base < 10 ? 10 : base;
    const cache = new Map<TrieNode, number>();
    let count = 0;
    const backBuffer = { last: '', count: 0 };

    function ref(n: number): string {
        return '#' + n.toString(radix) + ';';
    }

    function escape(s: string): string {
        return specialCharacters.has(s) ? ESCAPE + (specialCharacterMap.get(s) || s) : s;
    }

    function* flush() {
        while (backBuffer.count) {
            const n = Math.min(9, backBuffer.count);
            yield n > 1 ? backBuffer.last + n : backBuffer.last;
            backBuffer.last = BACK;
            backBuffer.count -= n;
        }
    }

    function* emit(s: string): Generator<string> {
        switch (s) {
            case EOW:
                yield* flush();
                backBuffer.last = EOW;
                backBuffer.count = 0;
                break;
            case BACK:
                backBuffer.count++;
                break;
            default:
                yield* flush();
                yield s;
        }
    }

    function* walk(node: TrieNode, depth: number): Generator<string> {
        const r = cache.get(node);
        if (r !== undefined) {
            yield* emit(ref(r));
            return;
        }
        if (node.c) {
            cache.set(node, count++);
            const c = [...node.c].sort((a, b) => (a[0] < b[0] ? -1 : 1));
            for (const [s, n] of c) {
                yield* emit(escape(s));
                yield* walk(n, depth + 1);
                yield* emit(BACK);
                if (depth === 0) yield* emit(EOL);
            }
        }
        // Output EOW after children so it can be optimized on read
        if (node.f) {
            yield* emit(EOW);
        }
    }

    function* serialize(node: TrieNode): Generator<string> {
        yield* walk(node, 0);
        yield* flush();
    }

    return generateHeader(radix, comment).concat(bufferLines(bufferLines(serialize(root), 120, '\n'), 10, ''));
}

function* toIterableIterator<T>(iter: Iterable<T>): IterableIterator<T> {
    yield* iter;
}

interface Stack {
    node: TrieNode;
    s: string;
}

interface ReduceResults {
    stack: Stack[];
    nodes: TrieNode[];
    root: TrieRoot;
    parser: Reducer | undefined;
}

type Reducer = (acc: ReduceResults, s: string) => ReduceResults;

export function importTrie(linesX: Iterable<string>): TrieRoot {
    const root: TrieRoot = trieNodeToRoot({}, {});

    let radix = 16;
    const comment = /^\s*#/;
    const iter = toIterableIterator(linesX);

    function parseHeaderRows(headerRows: string[]) {
        const header = headerRows.slice(0, 2).join('\n');
        const headerReg = /^TrieXv3\nbase=(\d+)$/;
        /* istanbul ignore if */
        if (!headerReg.test(header)) throw new Error('Unknown file format');
        radix = Number.parseInt(header.replace(headerReg, '$1'), 10);
    }

    function readHeader(iter: Iterator<string>) {
        const headerRows: string[] = [];
        // eslint-disable-next-line no-constant-condition
        while (true) {
            const next = iter.next();
            if (next.done) {
                break;
            }
            const line = next.value.trim().replace(/\r|\n/g, '');
            if (!line || comment.test(line)) {
                continue;
            }
            if (line === DATA) {
                break;
            }
            headerRows.push(line);
        }
        parseHeaderRows(headerRows);
    }

    readHeader(iter);

    const n = genSequence(iter)
        .concatMap((a) => a.split(''))
        .reduce(parseStream(radix), {
            nodes: [root],
            root,
            stack: [{ node: root, s: '' }],
            parser: undefined,
        });
    return n.root;
}

function parseStream(radix: number): Reducer {
    const eow: TrieNode = Object.freeze({ f: 1 });

    function parseReference(acc: ReduceResults, _: string): ReduceResults {
        let ref = '';

        function parser(acc: ReduceResults, s: string): ReduceResults {
            if (s === EOR) {
                const { root, nodes, stack } = acc;
                const r = parseInt(ref, radix);
                const top = stack[stack.length - 1];
                const p = stack[stack.length - 2].node;
                p.c?.set(top.s, nodes[r]);
                return { root, nodes, stack, parser: undefined };
            }
            ref = ref + s;
            return acc;
        }

        const { nodes } = acc;
        nodes.pop();
        return { ...acc, nodes, parser };
    }

    function parseEscapeCharacter(acc: ReduceResults, _: string): ReduceResults {
        let prev = '';
        const parser = function (acc: ReduceResults, s: string): ReduceResults {
            if (prev) {
                s = characterMap.get(prev + s) || s;
                return parseCharacter({ ...acc, parser: undefined }, s);
            }
            if (s === ESCAPE) {
                prev = s;
                return acc;
            }
            return parseCharacter({ ...acc, parser: undefined }, s);
        };
        return { ...acc, parser };
    }

    function parseCharacter(acc: ReduceResults, s: string): ReduceResults {
        const parser = undefined;
        const { root, nodes, stack } = acc;
        const top = stack[stack.length - 1];
        const node = top.node;
        node.c = node.c ?? new Map<string, TrieNode>();
        const n = { f: undefined, c: undefined, n: nodes.length };
        node.c.set(s, n);
        stack.push({ node: n, s });
        nodes.push(n);
        return { root, nodes, stack, parser };
    }

    function parseEOW(acc: ReduceResults, _: string): ReduceResults {
        const parser = parseBack;
        const { root, nodes, stack } = acc;
        const top = stack[stack.length - 1];
        const node = top.node;
        node.f = FLAG_WORD;
        if (!node.c) {
            top.node = eow;
            const p = stack[stack.length - 2].node;
            p.c?.set(top.s, eow);
            nodes.pop();
        }
        stack.pop();
        return { root, nodes, stack, parser };
    }

    const charactersBack = new Set((BACK + '23456789').split(''));
    function parseBack(acc: ReduceResults, s: string): ReduceResults {
        if (!charactersBack.has(s)) {
            return parserMain({ ...acc, parser: undefined }, s);
        }
        let n = s === BACK ? 1 : parseInt(s, 10) - 1;
        const { stack } = acc;
        while (n-- > 0) {
            stack.pop();
        }
        return { ...acc, parser: parseBack };
    }

    function parseIgnore(acc: ReduceResults, _: string): ReduceResults {
        return acc;
    }

    const parsers = new Map<string, Reducer>([
        [EOW, parseEOW],
        [BACK, parseBack],
        [REF, parseReference],
        [ESCAPE, parseEscapeCharacter],
        [EOL, parseIgnore],
        [LF, parseIgnore],
    ]);

    function parserMain(acc: ReduceResults, s: string): ReduceResults {
        const parser = acc.parser ?? parsers.get(s) ?? parseCharacter;
        return parser(acc, s);
    }
    return parserMain;
}
