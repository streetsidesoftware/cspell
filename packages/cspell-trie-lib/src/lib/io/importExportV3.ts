import { opAppend, pipe } from '@cspell/cspell-pipe/sync';

import type { TrieNode, TrieRoot } from '../TrieNode/TrieNode.js';
import { bufferLines } from '../utils/bufferLines.js';
import { BACK, EOL, EOR, EOW, ESCAPE, LF, REF } from './constants.js';
import { importTrieV3AsTrieRoot } from './importV3.js';

const specialCharacters = stringToCharSet(
    [EOW, BACK, EOL, REF, EOR, ESCAPE, LF, '0123456789', '`~!@#$%^&*()_-+=[]{};:\'"<>,./?\\|'].join(''),
);

const specialCharacterMap = new Map([
    ['\n', '\\n'],
    ['\r', '\\r'],
    ['\\', '\\\\'],
]);

const specialPrefix = stringToCharSet('~!');

const WORDS_PER_LINE = 20;

export const DATA = '__DATA__';

function generateHeader(base: number, comment: string): Iterable<string> {
    const header = ['#!/usr/bin/env cspell-trie reader', 'TrieXv3', 'base=' + base]
        .concat(comment ? comment.split('\n').map((a) => '# ' + a) : [])
        .concat(['# Data:', DATA]);
    return header.map((a) => a + '\n');
}

export interface ExportOptions {
    base?: number;
    comment?: string;
    /**
     * This will reduce the size of the `.trie` file by removing references to short suffixes.
     * But it does increase the size of the trie when loaded into memory.
     */
    optimizeSimpleReferences?: boolean;
    /**
     * To improve diffs, an EOL is added before each double letter prefix.
     * @default true
     */
    addLineBreaksToImproveDiffs?: boolean;
}

/**
 * Serialize a TrieRoot.
 */
export function serializeTrie(root: TrieRoot, options: ExportOptions | number = 16): Iterable<string> {
    options = typeof options === 'number' ? { base: options, addLineBreaksToImproveDiffs: false } : options;
    const { base = 16, comment = '', addLineBreaksToImproveDiffs: addBreaks = true } = options;
    const radix = base > 36 ? 36 : base < 10 ? 10 : base;
    const cache = new Map<TrieNode, number>();
    const cacheShouldRef = new Map<TrieNode, boolean>();
    let count = 0;
    const backBuffer = { last: '', count: 0, words: 0, eol: false };
    const optimizeSimpleReferences = options.optimizeSimpleReferences ?? false;
    const wordChars: string[] = [];

    function ref(n: number): string {
        return '#' + n.toString(radix) + ';';
    }

    function escape(s: string): string {
        return s in specialCharacters ? ESCAPE + (specialCharacterMap.get(s) || s) : s;
    }

    function* flush() {
        while (backBuffer.count) {
            const n = Math.min(9, backBuffer.count);
            yield n > 1 ? backBuffer.last + n : backBuffer.last;
            backBuffer.last = BACK;
            backBuffer.count -= n;
        }
        if (backBuffer.eol) {
            yield EOL;
            backBuffer.eol = false;
            backBuffer.words = 0;
        }
    }

    function* emit(s: string): Generator<string> {
        switch (s) {
            case EOW:
                yield* flush();
                backBuffer.last = EOW;
                backBuffer.count = 0;
                backBuffer.words++;
                break;
            case BACK:
                backBuffer.count++;
                break;
            case EOL:
                backBuffer.eol = true;
                break;
            default:
                if (backBuffer.words >= WORDS_PER_LINE) {
                    backBuffer.eol = true;
                }
                yield* flush();
                if (s.startsWith(REF)) {
                    backBuffer.words++;
                }
                yield s;
        }
    }

    function* walk(node: TrieNode, depth: number): Generator<string> {
        const r = cache.get(node);
        if (r !== undefined && (!optimizeSimpleReferences || !shouldSimpleRef(node))) {
            yield* emit(ref(r));
            return;
        }
        if (node.c) {
            if (addBreaks && depth > 0 && depth <= 2) {
                yield* emit(EOL);
            }
            cache.set(node, count++);
            const c = Object.entries(node.c).sort((a, b) => (a[0] < b[0] ? -1 : 1));
            for (const [s, n] of c) {
                wordChars[depth] = s;
                yield* emit(escape(s));
                yield* walk(n, depth + 1);
                yield* emit(BACK);
                if (depth === 0) yield* emit(EOL);
            }
        }
        // Output EOW after children so it can be optimized on read
        if (node.f) {
            yield* emit(EOW);
        }
        if (addBreaks && (depth === 2 || (depth === 3 && wordChars[0] in specialPrefix))) {
            yield* emit(EOL);
        }
    }

    function* serialize(node: TrieNode): Generator<string> {
        yield* walk(node, 0);
        yield* flush();
    }

    function _calcShouldSimpleRef(node: TrieNode): boolean {
        if (!node.c) return false;
        const values = Object.values(node.c);
        if (values.length !== 1) return false;
        const n = values[0];
        return !!n.f && (!n.c || !Object.values(n.c).length);
    }

    function shouldSimpleRef(node: TrieNode): boolean {
        const r = cacheShouldRef.get(node);
        if (r !== undefined) return r;
        const rr = _calcShouldSimpleRef(node);
        cacheShouldRef.set(node, rr);
        return rr;
    }

    return pipe(generateHeader(radix, comment), opAppend(bufferLines(serialize(root), 1200, '')));
}

interface Stack {
    node: TrieNode;
    s: string;
}

interface ReduceResults {
    stack: Stack[];
    nodes: TrieNode[];
    root: TrieRoot;
    parser: Reducer | undefined;
}

type Reducer = (acc: ReduceResults, s: string) => ReduceResults;

export function importTrie(srcLines: string[] | Iterable<string> | string): TrieRoot {
    const trie = importTrieV3AsTrieRoot(srcLines);
    return trie.root;
}

function stringToCharSet(values: string): Record<string, boolean | undefined> {
    const set: Record<string, boolean | undefined> = Object.create(null);
    const len = values.length;
    for (let i = 0; i < len; ++i) {
        set[values[i]] = true;
    }
    return set;
}
