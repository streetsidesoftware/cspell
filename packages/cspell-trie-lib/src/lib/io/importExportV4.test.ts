import { readFile, writeFile } from 'fs-extra';
import { genSequence } from 'gensequence';

import { resolveSample as resolveSamplePath } from '../../test/samples';
import { consolidate } from '../consolidate';
import * as Trie from '../index';
import type { TrieNode } from '../TrieNode';
import * as v3 from './importExportV3';
import { __testing__, importTrie, serializeTrie } from './importExportV4';

const sampleFile = resolveSamplePath('sampleV4.trie');

describe('Import/Export', () => {
    test('tests serialize / deserialize small sample', () => {
        const trie = Trie.buildTrie(smallSample).root;
        const expected = toTree(trie);
        const data = [...serializeTrie(trie, { base: 10, comment: 'Sample Words' })].join('');
        const root = importTrie(data);
        const words = [...Trie.iteratorTrieWords(root)];
        expect(words).toEqual([...smallSample].sort());
        const result = toTree(root);
        expect(result).toBe(expected);
    });

    test('tests serialize / deserialize specialCharacters', () => {
        const trie = Trie.buildTrie(specialCharacters).root;
        const data = [...serializeTrie(consolidate(trie), 10)];
        const root = importTrie(data);
        const words = [...Trie.iteratorTrieWords(root)];
        expect(words).toEqual([...specialCharacters].sort());
    });

    test('tests serialize / deserialize', async () => {
        const trie = Trie.buildTrie(sampleWords).root;
        const data = [
            ...serializeTrie(consolidate(trie), {
                base: 10,
                comment: 'Sample Words',
            }),
        ].join('');
        const root = importTrie(data);
        const words = [...Trie.iteratorTrieWords(root)];
        expect(words).toEqual([...sampleWords].sort());
        await writeFile(sampleFile, data);
    });

    test('tests deserialize from file', async () => {
        const sample = await readFile(sampleFile, 'utf8');
        const root = importTrie(sample);
        const words = [...Trie.iteratorTrieWords(root)];
        expect(words).toEqual([...sampleWords].sort());
    });

    test('tests serialize / deserialize trie', () => {
        const trie = Trie.buildTrie(sampleWords).root;
        const data = serializeTrie(trie, 16);
        const root = importTrie(data);
        const words = [...Trie.iteratorTrieWords(root)];
        expect(words).toEqual([...sampleWords].sort());
    });

    test('serialize / deserialize trie DAWG', () => {
        const trie = Trie.buildTrie(sampleWords).root;
        const trieDawg = consolidate(trie);
        const data = serializeTrie(trieDawg, 16);
        const root = importTrie(data);
        const words = [...Trie.iteratorTrieWords(root)];
        expect(words).toEqual([...sampleWords].sort());
    });

    test.each`
        options
        ${10}
        ${{ base: 10 }}
        ${{ base: 10, optimizeSimpleReferences: true }}
        ${{ base: 10, optimizeSimpleReferences: false }}
    `('serialize DAWG $options', ({ options }) => {
        const trie = Trie.createTriFromList(sampleWords);
        const trieDawg = consolidate(trie);
        const data = [...serializeTrie(trieDawg, options)];
        const root = importTrie(data);
        const words = [...Trie.iteratorTrieWords(root)];
        expect(words).toEqual([...sampleWords].sort());
        expect(data.join('')).toMatchSnapshot();
    });

    test.each`
        sampleWordList   | options
        ${'sample.txt'}  | ${{ addTrieBreaksToImproveDiffs: false }}
        ${'sample2.txt'} | ${{ addTrieBreaksToImproveDiffs: false }}
        ${'sample.txt'}  | ${{}}
        ${'sample2.txt'} | ${{}}
    `('Read sample and ensure results match $sampleWordList $options', async ({ sampleWordList, options }) => {
        const path = resolveSamplePath(sampleWordList);
        const content = await readFile(path, 'utf-8');
        const wordList = content
            .split('\n')
            .map((a) => a.trim())
            .filter((a) => !!a);

        const trie = Trie.buildTrie(wordList);
        wordList.sort();
        const data = [...serializeTrie(trie.root, options)].join('');
        const trie2 = importTrie(data);
        const wordsTrie = [...Trie.iteratorTrieWords(trie2)];
        expect(wordsTrie).toEqual(wordList);
        expect(data).toMatchSnapshot();
    });

    test.each`
        options
        ${10}
        ${{ base: 10 }}
        ${{ base: 10, optimizeSimpleReferences: true }}
        ${{ base: 10, optimizeSimpleReferences: false }}
        ${16}
        ${{ base: 16 }}
        ${{ base: 16, optimizeSimpleReferences: true }}
        ${{ base: 16, optimizeSimpleReferences: false }}
    `('serialize with V3 DAWG $options', ({ options }) => {
        const trie = Trie.createTriFromList(sampleWords);
        const trieDawg = consolidate(trie);
        const data = [...v3.serializeTrie(trieDawg, options)];
        const root = importTrie(data);
        const words = [...Trie.iteratorTrieWords(root)];
        expect(words).toEqual([...sampleWords].sort());
    });

    test('buildReferenceMap', () => {
        const trie = Trie.createTriFromList(sampleWords);
        const trieDawg = consolidate(trie);
        const refMap = __testing__.buildReferenceMap(trieDawg, 10);
        const counts = refMap.refCounts.map(([_, count]) => count);
        expect(counts.length).toBeGreaterThan(10);
    });
});

function toTree(root: TrieNode): string {
    function* walk(n: TrieNode, prefix: string): Generator<string> {
        const nextPrefix = '.'.repeat(prefix.length);
        if (n.c) {
            for (const c of [...n.c].sort((a, b) => (a[0] < b[0] ? -1 : 1))) {
                yield* walk(c[1], prefix + c[0]);
                prefix = nextPrefix;
            }
        }
        if (n.f) {
            yield prefix + '$\n';
        }
    }

    return ['\n', ...walk(root, '')].join('');
}

const specialCharacters = [
    'arrow <',
    'escape \\',
    'eol \n',
    'eow $',
    'ref #',
    'Numbers 0123456789',
    'Braces: {}[]()',
    'slash /',
    'a/b',
    'a/c',
];

const smallSample = genSequence(['lift', 'talk', 'walk', 'turn', 'burn', 'chalk', 'churn'])
    .concatMap(applyEndings)
    .toArray();

const sampleWords = [
    'journal',
    'journalism',
    'journalist',
    'journalistic',
    'journals',
    'journey',
    'journeyer',
    'journeyman',
    'journeymen',
    'joust',
    'jouster',
    'jousting',
    'jovial',
    'joviality',
    'jowl',
    'jowly',
    'joy',
    'joyful',
    'joyfuller',
    'joyfullest',
    'joyfulness',
    'joyless',
    'joylessness',
    'joyous',
    'joyousness',
    'joyridden',
    'joyride',
    'joyrider',
    'joyriding',
    'joyrode',
    'joystick',
    'Big Apple',
    'New York',
    'apple',
    'big apple',
    'fun journey',
    'long walk',
    'fun walk',
]
    .concat(specialCharacters)
    .concat(smallSample);

function applyEndings(s: string): string[] {
    const endings = ['', 'ed', 'er', 'ing', 's'];
    return endings.map((e) => s + e);
}
