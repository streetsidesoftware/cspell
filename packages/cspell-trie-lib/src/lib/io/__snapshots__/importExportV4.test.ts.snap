// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Import/Export serialize DAWG {"base": 10, "optimizeSimpleReferences": false} 1`] = `
"#!/usr/bin/env cspell-trie reader
TrieXv4
base=10
# 
# Data:
__DATA__
[
0,27,56,58,103,54,61,62,85,86,101
]
B
/* B */
i
/* Bi */
g Apple$7
<r
/* Br */
aces\\\\: \\\\{\\\\}\\\\[\\\\]\\\\(\\\\)$9<3
<2
N
/* N */
e
/* Ne */
w York$6
<u
/* Nu */
mbers \\\\0\\\\1\\\\2\\\\3\\\\4\\\\5\\\\6\\\\7\\\\8\\\\9$9<7
<2
a
/* a */
p#6<r
/* ar */
row \\\\<$5
<2
b
/* b */
i
/* bi */
g a#5<3
<u
/* bu */
rned$r$2ing$3s$$2
<2
c
/* c */
h
/* ch */
alk@2<3u@5<
<2
e
/* e */
o
/* eo */
l \\\\\\\\n$3w \\\\$$3
<s
/* es */
cape \\\\\\\\\\\\$6
<2
f
/* f */
u
/* fu */
n journey$7wal@1<5
<2
j
/* j */
o
/* jo */
urnalism$tic$2$3s$$2eyer$2man$2e@4<2$4ste@10<i@3<$3vialit@8<2$4wly$$2yfuller$st$4
ness$4$3lessn#120<$4ou#125<2ridde@4<2er$$i@3<3od#8<3stic@1<4$
<2
l
/* l */
i
/* li */
ft@2<2
<o
/* lo */
ng w@9<4
<2
r
/* r */
e
/* re */
f \\\\#$3
<2
t@6<
w
/* w */
a@7<2
"
`;

exports[`Import/Export serialize DAWG {"base": 10, "optimizeSimpleReferences": true} 1`] = `
"#!/usr/bin/env cspell-trie reader
TrieXv4
base=10
# 
# Data:
__DATA__
[
0,27,56,58,103,54,61,62,85,86,101
]
B
/* B */
i
/* Bi */
g Apple$7
<r
/* Br */
aces\\\\: \\\\{\\\\}\\\\[\\\\]\\\\(\\\\)$9<3
<2
N
/* N */
e
/* Ne */
w York$6
<u
/* Nu */
mbers \\\\0\\\\1\\\\2\\\\3\\\\4\\\\5\\\\6\\\\7\\\\8\\\\9$9<7
<2
a
/* a */
p#6<r
/* ar */
row \\\\<$5
<2
b
/* b */
i
/* bi */
g a#5<3
<u
/* bu */
rned$r$2ing$3s$$2
<2
c
/* c */
h
/* ch */
alk@2<3u@5<
<2
e
/* e */
o
/* eo */
l \\\\\\\\n$3w \\\\$$3
<s
/* es */
cape \\\\\\\\\\\\$6
<2
f
/* f */
u
/* fu */
n journey$7wal@1<5
<2
j
/* j */
o
/* jo */
urnalism$tic$2$3s$$2eyer$2man$2e@4<2$4ste@10<i@3<$3vialit@8<2$4wly$$2yfuller$st$4
ness$4$3lessn#120<$4ou#125<2ridde@4<2er$$i@3<3od#8<3stic@1<4$
<2
l
/* l */
i
/* li */
ft@2<2
<o
/* lo */
ng w@9<4
<2
r
/* r */
e
/* re */
f \\\\#$3
<2
t@6<
w
/* w */
a@7<2
"
`;

exports[`Import/Export serialize DAWG {"base": 10} 1`] = `
"#!/usr/bin/env cspell-trie reader
TrieXv4
base=10
# 
# Data:
__DATA__
[
0,27,56,58,103,54,61,62,85,86,101
]
B
/* B */
i
/* Bi */
g Apple$7
<r
/* Br */
aces\\\\: \\\\{\\\\}\\\\[\\\\]\\\\(\\\\)$9<3
<2
N
/* N */
e
/* Ne */
w York$6
<u
/* Nu */
mbers \\\\0\\\\1\\\\2\\\\3\\\\4\\\\5\\\\6\\\\7\\\\8\\\\9$9<7
<2
a
/* a */
p#6<r
/* ar */
row \\\\<$5
<2
b
/* b */
i
/* bi */
g a#5<3
<u
/* bu */
rned$r$2ing$3s$$2
<2
c
/* c */
h
/* ch */
alk@2<3u@5<
<2
e
/* e */
o
/* eo */
l \\\\\\\\n$3w \\\\$$3
<s
/* es */
cape \\\\\\\\\\\\$6
<2
f
/* f */
u
/* fu */
n journey$7wal@1<5
<2
j
/* j */
o
/* jo */
urnalism$tic$2$3s$$2eyer$2man$2e@4<2$4ste@10<i@3<$3vialit@8<2$4wly$$2yfuller$st$4
ness$4$3lessn#120<$4ou#125<2ridde@4<2er$$i@3<3od#8<3stic@1<4$
<2
l
/* l */
i
/* li */
ft@2<2
<o
/* lo */
ng w@9<4
<2
r
/* r */
e
/* re */
f \\\\#$3
<2
t@6<
w
/* w */
a@7<2
"
`;

exports[`Import/Export serialize DAWG 10 1`] = `
"#!/usr/bin/env cspell-trie reader
TrieXv4
base=10
# 
# Data:
__DATA__
[
0,27,56,58,103,54,61,62,85,86,101
]
B
/* B */
i
/* Bi */
g Apple$7
<r
/* Br */
aces\\\\: \\\\{\\\\}\\\\[\\\\]\\\\(\\\\)$9<3
<2
N
/* N */
e
/* Ne */
w York$6
<u
/* Nu */
mbers \\\\0\\\\1\\\\2\\\\3\\\\4\\\\5\\\\6\\\\7\\\\8\\\\9$9<7
<2
a
/* a */
p#6<r
/* ar */
row \\\\<$5
<2
b
/* b */
i
/* bi */
g a#5<3
<u
/* bu */
rned$r$2ing$3s$$2
<2
c
/* c */
h
/* ch */
alk@2<3u@5<
<2
e
/* e */
o
/* eo */
l \\\\\\\\n$3w \\\\$$3
<s
/* es */
cape \\\\\\\\\\\\$6
<2
f
/* f */
u
/* fu */
n journey$7wal@1<5
<2
j
/* j */
o
/* jo */
urnalism$tic$2$3s$$2eyer$2man$2e@4<2$4ste@10<i@3<$3vialit@8<2$4wly$$2yfuller$st$4
ness$4$3lessn#120<$4ou#125<2ridde@4<2er$$i@3<3od#8<3stic@1<4$
<2
l
/* l */
i
/* li */
ft@2<2
<o
/* lo */
ng w@9<4
<2
r
/* r */
e
/* re */
f \\\\#$3
<2
t@6<
w
/* w */
a@7<2
"
`;
